<!DOCTYPE html>
<html>
<head>
<title>RBPBench - Search Report</title>


<style>
    th, td {
        border: 1px solid black;
        padding: 8px;
        text-align: center;
    }
    th {
        background-color: #f2f2f2;
    }
    .page {
        page-break-after: always;
    }

    .title-container {
        display: flex;
        align-items: center;
    }
    .title-container img {
        margin-right: 10px; /* Adjust the spacing as needed */
    }

</style>

</head>

<div class="title-container">
    <img src="html_report_plots/logo.png" alt="Logo" width="175">
    <h1>Search Report</h1>
</div>


<body>
<p>List of available statistics and plots generated
by RBPBench (v1.0.7, rbpbench search):</p>
<ul>
<li><a href="#rbp-enrich-stats">RBP region score motif enrichment statistics</a></li>
<li><a href="#cooc-heat-map">RBP motif co-occurrences heat map</a></li>
<li><a href="#seq-len-plot">Input sequence length distribution</a></li>
<li><a href="#seqs-top-kmer-plot">Input sequences top k-mers plot</a></li>
<li><a href="#kmer-dist">Top vs bottom scoring regions k-mer distribution</a></li>
<li><a href="#seq-var-plot">Input sequences k-mer variation plot</a></li>
<li><a href="#input-seqs-kmer">Input sequences by k-mer content plot</a></li>
</ul>
<p>FIMO p-value threshold (--fimo-pval) = 0.001. # of considered input regions = 2. Region extension (upstream, downstream) = (0, 0).</p>
<p>&nbsp;</p>
<h2 id="rbp-enrich-stats">RBP region score motif enrichment statistics</h2>
<p><strong>Table:</strong> RBP region score motif enrichment statistics. Given a score for each genomic region (# input regions = 2), 
RBPbench checks whether motif-containing regions have significantly different scores compared to regions without motif hits.
Wilcoxon rank-sum test alternative hypothesis is set to 'greater', i.e., low p-values mean motif-containing regions have significantly higher scores.
In other words, a low test p-value for a given RBP indicates 
that higher-scoring regions are more likely to contain motif hits of the respective RBP.
<strong>NOTE</strong> that if scores associated with input genomic regions are all the same, p-values become meaningless 
(i.e., they result in p-values of 1.0).
Likewise, the p-value becomes non-informative if most or all input regions have RBP motif hits (i.e., very high hit region percentages).
By default, BED genomic regions input file column 5 is used as the score column (change with --bed-score-col).</p>
<table style="max-width: 800px; width: 100%; border-collapse: collapse; line-height: 0.8;">
<thead>
<tr>
<th>RBP ID</th>
<th># hit regions</th>
<th>% hit regions</th>
<th># motif hits</th>
<th>p-value</th>
</tr>
</thead>
<tbody>
<tr>
<td>PUM1</td>
<td>0</td>
<td>0.00</td>
<td>0</td>
<td>1.0</td>
</tr>
<tr>
<td>PUM2</td>
<td>0</td>
<td>0.00</td>
<td>0</td>
<td>1.0</td>
</tr>
</tbody>
</table>

<p>&nbsp;
&nbsp;</p>
<p>Column IDs have the following meanings: <strong>RBP ID</strong> -&gt; RBP ID from database or user-defined (typically RBP name), <strong># hit regions</strong> -&gt; number of input genomic regions with motif hits (after filtering and optional extension), <strong>% hit regions</strong> -&gt; percentage of hit regions over all regions (i.e., how many input regions contain &gt;= 1 RBP binding motif), <strong># motif hits</strong> -&gt; number of unique motif hits in input regions (removed double counts), <strong>p-value</strong> -&gt; Wilcoxon rank-sum test p-value.</p>
<p>&nbsp;</p>
<h2 id="cooc-heat-map">RBP motif co-occurrences heat map</h2>
<div>
<iframe src="html_report_plots/co-occurrence_plot.plotly.html" width="1200" height="1200"></iframe>
</div>

<p><strong>Figure:</strong> Heat map of motif hit co-occurrences (Fisher's exact test p-values) between RBPs. 
RBP motif hit co-occurrences that are not significant are colored white, while
significant co-occurrences are colored according to their -log10 p-value (used as legend color, i.e., the higher the more significant).
RBP co-occurrences with Benjamini-Hochberg multiple testing corrected p-values below 0.005 are considered significant. # of RBP co-occurrence comparisons: 1. # of significant co-occurrences: 0 (0.00%).
Fisher exact test alternative hypothesis is set to 'greater', i.e., significantly overrepresented co-occurrences are reported.
Hover box: 
<strong>1)</strong> RBP1.
<strong>2)</strong> RBP2.
<strong>3)</strong> p-value: Fisher's exact test p-value (calculated based on contingency table (6) between RBP1 and RBP2). 
<strong>4)</strong> p-value after filtering: p-value after filtering, i.e., p-value is kept if significant (&lt; 0.005), otherwise it is set to 1.0.
<strong>5)</strong> RBPs compaired. 
<strong>6)</strong> Counts[]: contingency table of co-occurrence counts (i.e., number of genomic regions with/without shared motif hits) between compaired RBPs, 
with format [[A, B], [C, D]], where 
A: RBP1 AND RBP2, 
B: NOT RBP1 AND RBP2
C: RBP1 AND NOT RBP2
D: NOT RBP1 AND NOT RBP2.
<strong>7)</strong> Mean minimum distance of RBP1 and RBP2 motifs (mean over all regions containing RBP1 + RBP2 motifs). Distances measured from motif center positions.
<strong>8)</strong> Over all regions containing RBP1 and RBP2 motif pairs, percentage of regions where RBP1 + RBP2 motifs are within 20 nt distance (set via --max-motif-dist).
<strong>9)</strong> Correlation: Pearson correlation coefficient between RBP1 and RBP2.
Genomic regions are labelled 1 or 0 (RBP motif present or not), resulting in a vector of 1s and 0s for each RBP.
Correlations are then calculated by comparing vectors for every pair of RBPs.
<strong>10)</strong> -log10 of p-value after filtering, used for legend coloring. Using p-value after filtering, all non-significant p-values become 0 
for easier distinction between significant and non-significant co-occurrences.</p>
<p>&nbsp;</p>
<h2 id="seq-len-plot">Input sequence length distribution</h2>
<div>
<iframe src="html_report_plots/seq_len_violin_plot.plotly.html" width="1000" height="1000"></iframe>
</div>

<p><strong>Figure:</strong> Input sequence lengths (after filtering and optional extension) violin plot.
Hover box over data points shows sequence ID, sequence (50 nt per line), sequence length, and motif hits. 
Motif hit format: motif ID, motif start - motif end.
Violin plot shows density distribution of sequence lengths, including min, max, median, 
and length quartiles (q1: 25th percentile, q3: 75th percentile).</p>
<p>&nbsp;</p>
<h2 id="seqs-top-kmer-plot">Input sequences top k-mers plot</h2>
<p><image src = "html_report_plots/seqs_top_kmers.bar_plot.png" width="1100px"></image></p>
<p><strong>Figure:</strong> Top 20 4-mers encountered in input sequences. 
In case of a uniform distribution with all 4-mers present, each 4-mer would have a percentage of 0.390625% (marked by red line).
Change k via --kmer-plot-k.</p>
<p>&nbsp;</p>
<h2 id="kmer-dist">Top vs bottom scoring regions k-mer distribution</h2>
<div>
<iframe src="html_report_plots/plotly_scatter_kmer.html" width="800" height="800"></iframe>
</div>

<p><strong>Figure:</strong> Sequence 4-mer percentages in the top 1 scoring and bottom 1 scoring input sites. In case of
a uniform distribution with all 4-mers present, each 4-mer would have a percentage of 0.390625%. R2 = 0.000250.
By default, BED genomic regions input file column 5 is used as the score column (change with --bed-score-col).
<strong>NOTE</strong> that this plot is only meaningful if the provided scores are related to binding site quality / binding affinity.
In this case, the plot can hint at RBP binding preferences (corresponding to enriched k-mers in the top input sites). 
Plot can be further modified by specifying top and bottom n scoring regions (--kmer-tb-plot-top-n, --kmer-tb-plot-bottom-n, 
by default top and bottom are split in half), or change k (--kmer-plot-k).</p>
<p>&nbsp;</p>
<h2 id="seq-var-plot">Input sequences k-mer variation plot</h2>
<div>
<iframe src="html_report_plots/seq_kmer_var_plot.plotly.html" width="1000" height="1000"></iframe>
</div>

<p><strong>Figure:</strong> k-mer variations in input sequences. Set k-mer size = 4 (change via --kmer-plot-k).
x-axis: k-mer variation, describing the variation of the k-mer over all input sequences (see coefficient of variation (CV) in hover box description).
y-axis: site %, i.e., percentage of sequences where k-mer is present.
By default, the correlation (Spearman correlation coefficient) between k-mer ratios and input region scores is used for coloring 
(alternatively use --seq-var-mode to change to k-mer %).
Hover box:
<strong>k-mer</strong> -&gt; observed k-mer.
<strong>Variation</strong> -&gt; coefficient of variation (CV) of the k-mer ratios over all input site sequences. 
CV is defined as the standard deviation of ratios divided by the mean ratio. 
Thus, the lower the CV of a k-mer, the more even its ratios over the input site sequences.
A CV of 0.0 for a given k-mer would mean that all sequences have exactly the same ratio of the k-mer.
In general, we can assume that k-mers with high CVs are less important for the RBP binding, since
they do not occur evenly over the dataset. In contrast, k-mers with low CVs are more 
evenly distributed over the dataset sequences, and thus should be more important for RBP binding 
(assuming a sufficient dataset quality + an affinity of the RBP towards specific k-mer sequences). 
<strong>k-mer #</strong> -&gt; number of times k-mer appears in all input sequences.
<strong>k-mer %</strong> -&gt; percentage of k-mer over all k-mers in the input sequences.
<strong>Site %</strong> -&gt; percentage of sequences where k-mer is present.
<strong>Correlation</strong> -&gt; correlation coefficient (Spearman) between k-mer ratios and input region scores.</p>
<p>&nbsp;</p>
<h2 id="input-seqs-kmer">Input sequences by k-mer content plot</h2>
<div>
<iframe src="html_report_plots/seqs_kmer_plot.plotly.html" width="1000" height="1000"></iframe>
</div>

<p><strong>Figure:</strong> Input sequences by k-mer content plot (k = 3, change via --kmer-pca-plot-k). 
Input sequences are represented by their k-mer content and visualized as 2D PCA plot.
The closer two sequences are, the more similar their k-mer content.
Apart from the k-mer features, sequence complexity (i.e., Shannon entropy) is used as an additional feature for each sequence.
Average mono-nucleotide percentages over all input sequences: A = 25.00%, C = 25.00%, G = 25.00%, T = 25.00%.</p>
<p>&nbsp;</p>
<script src="/home/polina/git/RBPBench_dev/rbpbench/content/sorttable.js" type="text/javascript"></script>
</body>

</html>

