<!DOCTYPE html>
<html lang="en">

<head>

    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="description" content="">
    <meta name="author" content="">

    <title>RCAS</title>
    <script>
        function showDiv() {
           document.getElementById('topbox').style.display = "block";
           $("input").hide();
        }
    </script>

</head>

<body>
    <nav class="navbar navbar-inverse">
      <div class="container-fluid">
        <div class="navbar-header">
          <a class="navbar-brand" href="#">RCAS</a>
        </div>
        <div>
          <ul class="nav navbar-nav">
            <li class="dropdown">
              <a class="dropdown-toggle" data-toggle="dropdown" href="#">Links
              <span class="caret"></span></a>
              <ul class="dropdown-menu">
                <li><a href="https://www.denbi.de/">de.NBI</a></li>
                <li><a href="https://www.mdc-berlin.de/">MDC</a></li>
                <li><a href="https://www.mdc-berlin.de/13800178/en/bimsb">BIMSB</a></li>
                <li><a href="https://www.mdc-berlin.de/20902467/en/research/core_facilities/cf_bioinformatic">Bioinformatics Platform</a></li>
              </ul>
            </li>
            <li class="dropdown">
              <a class="dropdown-toggle" data-toggle="dropdown" href="#">About the authors
              <span class="caret"></span></a>
              <ul class="dropdown-menu">
                <li><a href="http://bioinformatics.mdc-berlin.de/team.html#altuna-akalin-phd">Dr. Altuna Akalin</a></li>
                <li><a href="http://bioinformatics.mdc-berlin.de/team.html#dilmurat-yusuf-phd">Dr. Dilmurat Yusuf</a></li>
                <li><a href="http://bioinformatics.mdc-berlin.de/team.html#ricardo-wurmus">Ricardo Wurmus</a></li>
                <li><a href="http://bioinformatics.mdc-berlin.de/team.html#bora-uyar-phd">Dr. Bora Uyar</a></li>
              </ul>
            </li>
            <li><a href="https://github.com/BIMSBbioinfo/RCAS">GitHUB</a></li>
            <li><a href="http://dorina.mdc-berlin.de/">DoRiNA</a></li>
            <li><a href="http://bioinformatics.mdc-berlin.de/contact.html">Contact</a></li>
            <li><a href="#acknowledgements">Acknowledgements</a></li>
          </ul>
        </div>
      </div>
    </nav>

    <div class="container">
      <div class="jumbotron">
        <h1>The RNA Centric Annotation System Analysis Report <br> <br></h1>
      </div>
      <div class="row">
        <div class="col-sm-4">
          <h3>About RCAS</h3>
          <a href="http://bioinformatics.mdc-berlin.de/"> <img src="base/img/GroupLogo.png"> </a>
            <p> <br> <b>RCAS</b> (<i><b>R</b></i>NA <i><b>C</b></i>entric <i><b>A</b></i>nnotation <i><b>S</b></i>ystem) is an automated system that provides dynamic annotations for custom
              input files that contain transcriptomic target regions. Such transcriptomic target regions could be, for instance, peak regions detected by
              CLIP-Seq analysis that detect protein-RNA interactions, MeRIP-Seq analysis that detect RNA modifications (alias the epitranscriptome), or any
              collection of target regions at the level of the transcriptome. </p>

            <p id="topbox" style="display:none;">
              The input required for the RCAS is a collection of transcriptomic target regions in <a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1"<b>'BED' format</b>. </a>
              RCAS provides summary statistics and annotations according to the distribution of the genomic coordinates of the queried regions. Thus, the aim of RCAS is to
                provide the biological context and the means to quickly summarize, visualize and support hypothesis generation processes for high-throughput transcriptomic data.
            </p>
            <input id="more-button" type="button" value="Read More" onclick="showDiv()" />

        </div>
        <div class="col-sm-4">
          <h3>GO term enrichment analysis</h3>
          <a href="http://geneontology.org/page/go-citation-policy"> <img src="base/img/go-logo.png"> </a>
            <p> <br> <b>RCAS</b> overlays the input target regions with the annotated protein-coding genes and calculates the Gene Ontology (GO) terms that may be enriched or
              depleted in the input target regions compared to the background list of protein-coding genes. A Classical Fisher's Exact Test is applied for each GO term and the p-values
              obtained for each GO term is corrected for multiple testing using both the False Discovery Rate and the Family-Wise Error Rate.
            </p>

        </div>
        <div class="col-sm-4">
          <h3>Gene Set Enrichment Analysis </h3>
            <a href="http://software.broadinstitute.org/gsea/msigdb"> <img src="base/img/MSigDB-logo.gif"> </a>
            <p> <br> Similarly to the GO term enrichment analysis, <b>RCAS</b> also detects sets of genes as annotated in the Molecular Signatures
              Database that are enriched or depleted in the queried target regions. Results are corrected for multiple-testing according to both the False Discovery Rate and the Family-Wise Error Rate.
            </p>
        </div>
      </div>
    </div>


</body>

</html>
