<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="Content-Type" content="text/html;charset=utf-8" >
    <style type="text/css">
        body {
            padding-bottom: 30px;
        }
        .content {
            margin-left: 120px;
            padding-top: 50px;
            clear: both;
        }
        #header {
            display: block;
            width: 100%;
            margin-bottom: 1em;
        }
        /*#report_date { *//* not used in Quast html report */
            /*margin-top: -13px;*/
            /*margin-bottom: 20px;*/
        /*}*/
    </style>

    <link rel="stylesheet" type="text/css" href="report_html_aux/bootstrap/bootstrap.min.css" />
    <link rel="stylesheet" type="text/css" href="report_html_aux/common.css" />
    <link rel="stylesheet" type="text/css" href="report_html_aux/report.css" />

    <script type="text/javascript" src="report_html_aux/jquery-1.8.2.min.js"></script>
    <script type="text/javascript" src="report_html_aux/flot/jquery.flot.min.js"></script>
    <script type="text/javascript" src="report_html_aux/flot/jquery.flot.dashes.js"></script>
    <script type="text/javascript" src="report_html_aux/bootstrap/bootstrap.min.js"></script>
    <script type="text/javascript" src="report_html_aux/bootstrap/bootstrap-tooltip-vlad.js"></script>

    <script type="text/javascript" src="report_html_aux/scripts/hsvToRgb.js"></script>
    <script type="text/javascript" src="report_html_aux/scripts/utils.js"></script>
    <script type="text/javascript" src="report_html_aux/scripts/build_total_report_common.js"></script>
    <script type="text/javascript" src="report_html_aux/scripts/build_total_report.js"></script>
    <script type="text/javascript" src="report_html_aux/scripts/draw_cumulative_plot.js"></script>
    <script type="text/javascript" src="report_html_aux/scripts/draw_nx_plot.js"></script>
    <script type="text/javascript" src="report_html_aux/scripts/draw_gc_plot.js"></script>
    <script type="text/javascript" src="report_html_aux/scripts/draw_genes_plot.js"></script>
    <script type="text/javascript" src="report_html_aux/scripts/build_report_common.js"></script>
    <script type="text/javascript" src="report_html_aux/scripts/build_report.js"></script>
    <script type="text/javascript" src="report_html_aux/dragtable.js"></script>

    <script type="text/javascript">
        $(function() {
            buildReport();

            dragTable.init();

            $('#contigs_are_ordered').show();

            $("[rel=tooltip]").tooltip({ animation: false, });
        });
    </script>
</head>

<body>
<div class="content">
    <div id='header'>
        <h1 style='font-size: 1.5em;'><span class='smallcapitals' id='quast_name'>QUAST</span>
            <span id='report_name'>report</span></h1>
        <p id='report_date'></p>
        <p id='mincontig'>Loading...</p>
        <p id='per_ref_msg'></p>

        <p id="data_set_p">
            <span id="main_ref_genome">
                <span style="display: none;" id="reference_length">
                    Reference size:
                    <span class='val'></span><span class='rhs'>&nbsp;</span>bp,
                </span>

                <span style="display: none;" id="reference_gc">G+C content:
                    <span class='val'></span><span class='rhs'>&nbsp;</span>%
                </span>

                <span style="display: none;" id="reference_genes"><br>
                    <span class='val'></span> genes
                </span>

                <span style="display: none;" id="reference_operons">,
                    <span class='val'></span> operons
                </span>
            </span>
        </p>
    </div>

    <noscript>
        <div style="margin-left: 0;">
            Unfortunately, <span style="font-weight: bold">JavaScript</span> in your
            browser <span style="font-weight: bold">is disabled</span> or is not supported.</br>
            We need JavaScript to build report and plots.
        </div>
    </noscript>

    <div class='report-and-plots'>
        <div class='report' id='main_report'>
        </div>

        <div class='plots'>
            <table style='margin: -3px;'>
                <tr id='all-switches'>
                    <td>
                        <div id='plots-switches'></div>
                    </td>
                    <td style='text-align: right; padding-right: 22px;'>
                        <span id='scale-placeholder'></span>
                    </td>
                </tr>
                <tr>
                    <td colspan="2">
                        <div id='plot-placeholder' style=''></div>
                    </td>
                    <td>
                        <div id='legend-placeholder'></div>
                    </td>
                </tr>
                <tr>
                    <td>
                        <p id="contigs_are_ordered">Contigs are ordered from largest (contig #1) to smallest.</p>
                        <p id="gc_info">Contigs are broken into nonoverlapping 100 bp windows.
                            Plot shows numbers of windows for each GC percentage.</p>
                    </td>
                </tr>
            </table>
        </div>

        <div style="clear: left"></div>
        <div style="clear: right"></div>

        <div class='json-code'>
            <div id='total-report-json'>
                {"minContig":500,"report":[["Statistics without reference",[{"values":[20,12],"quality":"Less is better","isMain":true,"metricName":"# contigs"},{"values":[47,35],"quality":"Less is better","isMain":false,"metricName":"# contigs (>= 0 bp)"},{"values":[11,7],"quality":"Less is better","isMain":false,"metricName":"# contigs (>= 1000 bp)"},{"values":[55106,55421],"quality":"More is better","isMain":true,"metricName":"Largest contig"},{"values":[180875,173140],"quality":"More is better","isMain":true,"metricName":"Total length"},{"values":[187773,179095],"quality":"More is better","isMain":false,"metricName":"Total length (>= 0 bp)"},{"values":[173623,169301],"quality":"More is better","isMain":false,"metricName":"Total length (>= 1000 bp)"},{"values":[49658,49658],"quality":"More is better","isMain":true,"metricName":"N50"},{"values":[17535,27260],"quality":"More is better","isMain":false,"metricName":"N75"},{"values":[2,2],"quality":"Less is better","isMain":false,"metricName":"L50"},{"values":[4,3],"quality":"Less is better","isMain":false,"metricName":"L75"},{"values":["45.35","45.15"],"quality":"Equal","isMain":false,"metricName":"GC (%)"}]],["Misassemblies",[{"values":[0,0],"quality":"Less is better","isMain":true,"metricName":"# misassemblies"},{"values":[0,0],"quality":"Less is better","isMain":false,"metricName":"    # relocations"},{"values":[0,0],"quality":"Less is better","isMain":false,"metricName":"    # translocations"},{"values":[0,0],"quality":"Less is better","isMain":false,"metricName":"    # inversions"},{"values":[0,0],"quality":"Less is better","isMain":false,"metricName":"    # interspecies translocations"},{"values":[0,0],"quality":"Less is better","isMain":false,"metricName":"# possibly misassembled contigs"},{"values":[0,0],"quality":"Less is better","isMain":false,"metricName":"# misassembled contigs"},{"values":[0,0],"quality":"Less is better","isMain":true,"metricName":"Misassembled contigs length"},{"values":[0,0],"quality":"Less is better","isMain":false,"metricName":"# local misassemblies"}]],["Unaligned",[{"values":[4,3],"quality":"Less is better","isMain":false,"metricName":"# fully unaligned contigs"},{"values":[53102,51257],"quality":"Less is better","isMain":false,"metricName":"Fully unaligned length"},{"values":[0,0],"quality":"Less is better","isMain":false,"metricName":"# partially unaligned contigs"},{"values":[0,0],"quality":"Less is better","isMain":false,"metricName":"    # with misassembly"},{"values":[0,0],"quality":"Less is better","isMain":false,"metricName":"    # both parts are significant"},{"values":[0,0],"quality":"Less is better","isMain":false,"metricName":"Partially unaligned length"}]],["Mismatches",[{"values":[44,34],"quality":"Less is better","isMain":false,"metricName":"# mismatches"},{"values":[2,0],"quality":"Less is better","isMain":false,"metricName":"# indels"},{"values":[2,0],"quality":"Less is better","isMain":false,"metricName":"Indels length"},{"values":["35.92","27.77"],"quality":"Less is better","isMain":true,"metricName":"# mismatches per 100 kbp"},{"values":["1.63","0.00"],"quality":"Less is better","isMain":true,"metricName":"# indels per 100 kbp"},{"values":[2,0],"quality":"Less is better","isMain":false,"metricName":"    # short indels"},{"values":[0,0],"quality":"Less is better","isMain":false,"metricName":"    # long indels"},{"values":[14,0],"quality":"Less is better","isMain":false,"metricName":"# N's"},{"values":["7.74","0.00"],"quality":"Less is better","isMain":true,"metricName":"# N's per 100 kbp"}]],["Genome statistics",[{"values":["98.002","97.942"],"quality":"More is better","isMain":true,"metricName":"Genome fraction (%)"},{"values":["1.043","1.002"],"quality":"Less is better","isMain":true,"metricName":"Duplication ratio"},{"values":[55106,55421],"quality":"More is better","isMain":false,"metricName":"Largest alignment"},{"values":[49658,49658],"quality":"More is better","isMain":false,"metricName":"NG50"},{"values":[49658,49658],"quality":"More is better","isMain":false,"metricName":"NG75"},{"values":[17534,16099],"quality":"More is better","isMain":false,"metricName":"NA50"},{"values":[27969,27260],"quality":"More is better","isMain":true,"metricName":"NGA50"},{"values":[17534,16099],"quality":"More is better","isMain":false,"metricName":"NGA75"},{"values":[2,2],"quality":"Less is better","isMain":false,"metricName":"LG50"},{"values":[2,2],"quality":"Less is better","isMain":false,"metricName":"LG75"},{"values":[3,3],"quality":"Less is better","isMain":false,"metricName":"LA50"},{"values":[2,2],"quality":"Less is better","isMain":false,"metricName":"LGA50"},{"values":[3,3],"quality":"Less is better","isMain":false,"metricName":"LGA75"}]],["Predicted genes",[]],["Reference statistics",[{"values":[124998,124998],"quality":"Equal","isMain":false,"metricName":"Reference length"},{"values":["45.30","45.30"],"quality":"Equal","isMain":false,"metricName":"Reference GC (%)"}]]],"referenceName":"combined_reference","date":"18 August 2015, Tuesday, 17:45:21","order":[0,1],"assembliesNames":["dataset_3.dat","dataset_4.dat"]}
            </div>
            <div id='qualities-json'>
                {{ qualities }}
            </div>
            <div id='main-metrics-json'>
                {{ mainMetrics }}
            </div>

            <div id="contigs-lengths-json">
                {"lists_of_lengths":[[55106,49658,27969,17535,11102,3898,3427,1610,1141,1120,1057,987,929,862,850,832,747,714,694,637],[55421,49658,27260,16099,10428,6988,3447,885,817,729,714,694]],"filenames":["dataset_3.dat","dataset_4.dat"]}
            </div>
            <div id='aligned-contigs-lengths-json'>
                {"lists_of_lengths":[[55106,27969,17534,11101,3898,3427,1141,1057,987,929,862,850,832,747,694,637],[55421,27260,16099,10428,6988,3447,817,729,694]],"filenames":["dataset_3.dat","dataset_4.dat"]}
            </div>
            <div id='assemblies-lengths-json'>
                {"assemblies_lengths":[180875,173140],"filenames":["dataset_3.dat","dataset_4.dat"]}
            </div>
            <div id='reference-length-json'>
                {"reflen":124998}
            </div>
            <div id='genes-in-contigs-json'>
                {{ genesInContigs }}
            </div>
            <div id='operons-in-contigs-json'>
                {{ operonsInContigs }}
            </div>
            <div id='gc-json'>
                {"list_of_GC_distributions":[[[0.0,1.0,2.0,3.0,4.0,5.0,6.0,7.0,8.0,9.0,10.0,11.0,12.0,13.0,14.0,15.0,16.0,17.0,18.0,19.0,20.0,21.0,22.0,23.0,24.0,25.0,26.0,27.0,28.0,29.0,30.0,31.0,32.0,33.0,34.0,35.0,36.0,37.0,38.0,39.0,40.0,41.0,42.0,43.0,44.0,45.0,46.0,47.0,48.0,49.0,50.0,51.0,52.0,53.0,54.0,55.0,56.0,57.0,58.0,59.0,60.0,61.0,62.0,63.0,64.0,65.0,66.0,67.0,68.0,69.0,70.0,71.0,72.0,73.0,74.0,75.0,76.0,77.0,78.0,79.0,80.0,81.0,82.0,83.0,84.0,85.0,86.0,87.0,88.0,89.0,90.0,91.0,92.0,93.0,94.0,95.0,96.0,97.0,98.0,99.0,100.0],[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,2,4,5,3,4,5,6,14,13,24,33,36,29,43,42,57,71,60,74,78,89,90,78,86,79,81,64,65,63,60,67,52,52,52,43,44,27,20,22,22,23,9,8,4,3,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]],[[0.0,1.0,2.0,3.0,4.0,5.0,6.0,7.0,8.0,9.0,10.0,11.0,12.0,13.0,14.0,15.0,16.0,17.0,18.0,19.0,20.0,21.0,22.0,23.0,24.0,25.0,26.0,27.0,28.0,29.0,30.0,31.0,32.0,33.0,34.0,35.0,36.0,37.0,38.0,39.0,40.0,41.0,42.0,43.0,44.0,45.0,46.0,47.0,48.0,49.0,50.0,51.0,52.0,53.0,54.0,55.0,56.0,57.0,58.0,59.0,60.0,61.0,62.0,63.0,64.0,65.0,66.0,67.0,68.0,69.0,70.0,71.0,72.0,73.0,74.0,75.0,76.0,77.0,78.0,79.0,80.0,81.0,82.0,83.0,84.0,85.0,86.0,87.0,88.0,89.0,90.0,91.0,92.0,93.0,94.0,95.0,96.0,97.0,98.0,99.0,100.0],[9,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,3,0,1,5,2,5,7,3,23,11,21,23,28,30,40,42,61,64,66,59,77,80,85,77,83,74,58,79,75,57,63,65,48,36,50,40,43,36,23,21,16,12,14,5,6,2,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0]]],"lists_of_gc_info":null,"filenames":["dataset_3.dat","dataset_4.dat"]}
            </div>
        </div>
    </div>

    <div class='json-code'>
        <div id='glossary-json'>
            {
    "# contigs"                     : "<span>is the total number of contigs in the assembly.</span>",
    "Largest contig"                : "<span>is the length of the longest contig in the assembly.</span>",
    "Total length"                  : "<span>is the total number of bases in the assembly.</span>",
    "Reference length"              : "<span>is the total number of bases in the reference.</span>",

    "N50"                           : "<span>is the contig length such that using longer or equal length contigs produces half (50%) of the bases of the assembly. Usually there is no value that produces exactly 50%, so the technical definition is the maximum length x such that using contigs of length at least x accounts for at least 50% of the total assembly length.</span>",
    "NG50"                          : "<span>is the contig length such that using longer or equal length contigs produces half (50%) of the bases of the reference genome. This metric is computed only if a reference genome is provided.</span>",
    "N75"                           : "<span>is the contig length such that using longer or equal length contigs produces 75% of the bases of the assembly. Usually there is no value that produces exactly 75%, so the technical definition is the maximum length x such that using contigs of length at least x accounts for at least 75% of the total assembly length.</span>",
    "NG75"                          : "<span>is the contig length such that using longer or equal length contigs produces 75% of the bases of the reference genome. This metric is computed only if a reference genome is provided.</span>",

    "L50"                           : "<span>is the minimum number of contigs that produce half (50%) of the bases of the assembly. In other words, it's the number of contigs of length at least N50.</span>",
    "LG50"                          : "<span>is the minimum number of contigs that produce half (50%) of the bases of the reference genome. In other words, it's the number of contigs of length at least NG50. This metric is computed only if a reference genome is provided.</span>",
    "L75"                           : "<span>is the minimum number of contigs that produce 75% of the bases of the assembly. In other words, it's the number of contigs of length at least N75.</span>",
    "LG75"                          : "<span>is the minimum number of contigs that produce 75% of the bases of the reference genome. In other words, it's the number of contigs of length at least NG75. This metric is computed only if a reference genome is provided.</span>",

    "NA50"                          : "<span>is N50 where the lengths of aligned blocks are counted instead of contig lengths. I.e., if a contig has a misassembly with respect to the reference, the contig is broken into smaller pieces. This metric is computed only if a reference genome is provided.</span>",
    "NGA50"                         : "<span>is NG50 where the lengths of aligned blocks are counted instead of contig lengths. I.e., if a contig has a misassembly with respect to the reference, the contig is broken into smaller pieces. This metric is computed only if a reference genome is provided.</span>",
    "NA75"                          : "<span>is N75 where the lengths of aligned blocks are counted instead of contig lengths. I.e., if a contig has a misassembly with respect to the reference, the contig is broken into smaller pieces. This metric is computed only if a reference genome is provided.</span>",
    "NGA75"                         : "<span>is NG75 where the lengths of aligned blocks are counted instead of contig lengths. I.e., if a contig has a misassembly with respect to the reference, the contig is broken into smaller pieces. This metric is computed only if a reference genome is provided.</span>",

    "LA50"                          : "<span>is L50 where aligned blocks are counted instead of contigs. I.e., if a contig has a misassembly with respect to the reference, the contig is broken into smaller pieces.</span>",
    "LGA50"                         : "<span>is LG50 where aligned blocks are counted instead of contigs. I.e., if a contig has a misassembly with respect to the reference, the contig is broken into smaller pieces.</span>",
    "LA75"                          : "<span>is L75 where aligned blocks are counted instead of contigs. I.e., if a contig has a misassembly with respect to the reference, the contig is broken into smaller pieces.</span>",
    "LGA75"                         : "<span>is LG75 where aligned blocks are counted instead of contigs. I.e., if a contig has a misassembly with respect to the reference, the contig is broken into smaller pieces.</span>",

    "Average %IDY"                  : "<span>is the average of alignment identity percent (Nucmer measure of alignment accuracy) among all contigs.</span>",
    "# misassemblies"               : "<span>is the number of positions in the assembled contigs where the left flanking sequence aligns over 1 kbp away from the right flanking sequence on the reference (<i>relocation</i>) or they overlap on more than 1 kbp (<i>relocation</i>) or flanking sequences align on different strands (<i>inversion</i>) or different chromosomes (<i>translocation</i>).</span>",
    "# misassembled contigs"        : "<span>is the number of contigs that contain misassembly events.</span>",
    "Misassembled contigs length"   : "<span>is the number of total bases contained in all contigs that have one or more misassemblies.</span>",

    "# relocations"                 : "<span>is the number of relocation events among all misassembly events. Relocation is a misassembly where the left flanking sequence aligns over 1 kbp away from the right flanking sequence on the reference, or they overlap by more than 1 kbp and both flanking sequences align on the same chromosome.</span>",
    "# translocations"              : "<span>is the number of translocation events among all misassembly events. Translocation is a misassembly where the flanking sequences align on different chromosomes.</span>",
    "# interspecies translocations" : "<span>is the number of interspecies translocation events among all misassembly events. Interspecies translocation is a misassembly where the flanking sequences align on different references (--meta only).</span>",
    "# inversions"                  : "<span>is the number of inversion events among all misassembly events. Inversion is a misassembly where it is not a <i>relocation</i> and the flanking sequences align on opposite strands of the same chromosome.</span>",
    "# local misassemblies"         : "<span>is the number of local misassemblies. We define a local misassembly breakpoint as a breakpoint that satisfies these conditions: <ol> <li>Two or more distinct alignments cover the breakpoint.</li> <li>The gap between left and right flanking sequences is less than 1 kbp.</li> <li>The left and right flanking sequences both are on the same strand of the same chromosome of the reference genome.</li> </ol> </span>",
    "# possibly misassembled contigs": "<span>is the number of contigs that contain large unaligned fragment and thus could possibly contain interspecies translocation with unknown reference.</span>",

    "# fully unaligned contigs"     : "<span>is the number of contigs that have no alignment to the reference sequence.</span>",
    "Fully unaligned length"        : "<span>is the total number of bases contained in all fully unaligned contigs.</span>",
    "# partially unaligned contigs" : "<span>is the number of contigs that are not fully unaligned ones but have fragments with no alignment to the reference.</span>",
    "# with misassembly"            : "<span>is the number of partially unaligned contigs that contain misassembly events in their aligned fragment. Note that such misassemblies are not counted in <i># misassemblies</i> and other <i>misassemblies</i> statistics.</span>",
    "# both parts are significant"  : "<span>is the number of partially unaligned contigs that contain both aligned and unaligned fragments of length &ge; <i>min-contig threshold</i>.</span>",
    "Partially unaligned length"    : "<span>is the total number of unaligned bases in all partially unaligned contigs.</span>",

    "# ambiguous contigs"           : "<span>is the number of contigs that have reference alignments of equal quality in multiple locations on the reference.</span>",
    "Ambiguous contigs length"      : "<span>is the total number of bases contained in all ambiguous contigs.</span>",

    "Genome fraction (%)"           : "<span>is the total number of aligned bases in the reference, divided by the genome size. A base in the reference genome is counted as aligned if there is at least one contig with at least one alignment to this base. Contigs from repeat regions may map to multiple places, and thus may be counted multiple times in this quantity.</span>",
    "GC (%)"                        : "<span>is the total number of G and C nucleotides in the assembly, divided by the total length of the assembly.</span>",
    "Reference GC (%)"              : "<span>is the total number of G and C nucleotides in the reference, divided by the total length of the reference.</span>",

    "# mismatches per 100 kbp"      : "<span>is the average number of mismatches per 100000 aligned bases.</span>",
    "# mismatches"                  : "<span>is the number of mismatches in all aligned bases.</span>",
    "# indels per 100 kbp"          : "<span>is the average number of indels per 100000 aligned bases.</span>",
    "# indels"                      : "<span>is the number of indels in all aligned bases</span>",
    "# short indels"                : "<span>is the number of indels of length less or equal to 5 bp</span>",
    "# long indels"                 : "<span>is the number of indels of length greater than 5 bp</span>",
    "Indels length"                 : "<span>is the number of total bases contained in all indels</span>",

    "# genes"                       : "<span>is the number of genes in the assembly (complete and partial), based on a user-provided annotated list of gene positions in the reference genome. A gene counts as 'partially covered' if the assembly contains at least 100 bp of this gene but not the whole gene.</span>",
    "# operons"                     : "<span>is the number of operons in the assembly (complete and partial), based on a user-provided annotated list of operon positions in the reference genome. An operon counts as 'partially covered' if the assembly contains at least 100 bp of this operon but not the whole operon.</span>",

    "Cumulative length"             : "<span>plot shows the growth of assembly contig lengths. On the x-axis, contigs are ordered from largest (contig #1) to smallest. The y-axis gives the size of the x largest contigs in the assembly.</span>",
    "Nx"                            : "<span>plot shows the Nx metric value as x varies from 0 to 100.</span>  <span>Nx is the minimum contig length <b>y</b> such that using contigs of length at least <b>y</b> accounts for at least x% of the total assembly length.</span>",
    "NGx"                           : "<span>plot shows the NGx metric value as x varies from 0 to 100.</span> <span>NGx is the minimum contig length <b>y</b> such that using contigs of length at least <b>y</b> accounts for at least x% of the bases of the reference genome. This metric is computed only if a reference genome is provided.</span>",
    "NAx"                           : "<span>plot shows the NAx metric value as x varies from 0 to 100.</span> <span>NAx is computed similarly to Nx, but based on lengths of aligned blocks instead of contig lengths. Contigs are broken into aligned blocks at misassembly breakpoints. NAx is the minimum block length <b>y</b> such that using blocks of length at least <b>y</b> accounts for at least x% of the bases of the assembly. This metric is computed only if a reference genome is provided.</span>",
    "NGAx"                          : "<span>plot shows the NGAx metric value as x varies from 0 to 100.</span><span>NGAx is computed similarly to NGx, but based on lengths of aligned blocks instead of contig lengths. Contigs are broken at misassembly breakpoints. NGAx is the minimum block length <b>y</b> such that using blocks of length at least <b>y</b> accounts for at least x% of the bases of the reference genome. This metric is computed only if a reference genome is provided.</span>",
    "GC content"                    : "<span>plot shows the distribution of GC percentage among the contigs, i.e., the total number of bases in contigs with such GC content. Typically, the distribution is approximately Gaussian. However, for some genomes it is not Gaussian. For assembly projects with contaminants, the GC distribution of the contaminants often differs from the reference genome and may give a superposition of multiple curves with different peaks.</span>",

    "Duplication ratio"             : "<span>is the total number of aligned bases in the assembly (i.e. <i>Total length</i> - <i>Fully unaligned length</i> - <i>Partially unaligned length</i>), divided by the total number of aligned bases in the reference (see the <b>Genome fraction (%)</b> metric). If the assembly contains many contigs that cover the same regions of the reference, its <i>Duplication ratio</i> may be much larger than 1. This may occur due to overestimating repeat multiplicities and due to small overlaps between contigs, among other reasons.</span>",
    "Largest alignment"             : "<span>is the length of the largest continuous alignment in the assembly. This metric is always equal to the <i>Largest contig</i> metric but it can be smaller if the largest contig of the assembly contains a misassembly event.</span>",

    "# N's"                         : "<span>is the total number of uncalled bases (N's) in the assembly.</span>",
    "# N's per 100 kbp"             : "<span>is the average number of uncalled bases (N's) per 100000 assembly bases.</span>"
}

        </div>
    </div>
</div>
</body>
</html>
